# Corretor Ortográfico (NLP)

Desenvolvimento de um corretor ortográfico utilizando Python.

# 1. - Importar arquivo e leitura dos dados

In [2]:
# Abrindo o arquivo txt, armazenar seu conteúdo em uma variável declarada.

with open('C:\\Users\\batis\\Documents\\Jupyter_Notebook\\treinamento.txt', encoding="utf8",mode='r') as f:
    treinamento = f.read()
    
# Lendo arquivo e imprimindo o conteúdo
print(treinamento[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


Depois de fazer a leitura do arquivo temos que fazer a "tokenização" que é a separação do texto em "tokens".

**Definição de Token:**
    - Token é uma sequência de caracteres, separados por um limitador, que pode ser um espaço em branco, pontuação ou quebra de linhas, por exemplo.


In [3]:
# Mostrando quantos caracteres possui o arquivo
len(treinamento)

2605046

# 2. - Fatiando o conteúdo do arquivo

Utilizando o método _**split()**_ podemos identificar quantas palavras compoem o arquivo e não a quantidade de caracteres

In [4]:
# Exemplo da utilização do método
exemplo_texto = 'Olá Mundo!! Terra falando.'

# Tokenizando o texto com o split()
tokens = exemplo_texto.split()

# Imprimindo o resultado, ele separa as palavras pelo 'espaço' entre elas
print(tokens)

# Contando o tamanho da lista gerada
len(tokens) 

['Olá', 'Mundo!!', 'Terra', 'falando.']


4

**Import lib nltk**

In [5]:
# Import lib e baixando o pacote punkt
import nltk

nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\batis\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [6]:
# Tokenizando o texto utilizando a lib nltk e mostrando o resultado

palavras_separadas = nltk.tokenize.word_tokenize(exemplo_texto)
print(palavras_separadas)
len(palavras_separadas)

['Olá', 'Mundo', '!', '!', 'Terra', 'falando', '.']


7

### Nota/Observação

Notamos que ao utilizar a biblioteca nltk, é possível separar os caracteres (seja caractere especial ou pontuação) das palavras encontradas, porém ainda fazem parte de uma mesma lista.

# 2.1 - Separando os dados

Para realizar a separação utilizo o método isalpha()

In [7]:
'/.'.isalpha()

False

In [8]:
'ç'.isalpha()

True

In [9]:
'à'.isalpha()

True

### Nota/Observação
Este método consiste em analisar e verificar se em uma sequência existe apenas letras.

In [10]:
# Criando uma função para separar as palavras
def separa_palavras(lista_tokens):
    
    # Criando uma lista vázia
    lista_palavras = []
    
    for token in lista_tokens:
        
        if token.isalpha():
            
            lista_palavras.append(token)
    
    return lista_palavras

# Chamando a função e passando o parâmetro
separa_palavras(palavras_separadas)

['Olá', 'Mundo', 'Terra', 'falando']

Função criada, vamos para o próximo passo, criar o nosso corpus

In [11]:
lista_tokens = nltk.tokenize.word_tokenize(treinamento)

# Chamando a função e passando o corpus como parâmetro
lista_palavras = separa_palavras(lista_tokens)

# Imprimindo o resultado
print(f'O número total de palavras em nosso corpus é de : {len(lista_palavras)}')

O número total de palavras em nosso corpus é de : 403104


In [12]:
# Amostragem dos resultados
print(lista_palavras[:5])

['imagem', 'Temos', 'a', 'seguinte', 'classe']


# 3. - Normalização

Normalizar as palavras é passar todas para um mesmo padrão, como por exemplo deixar todas em letras minúsculas.

In [13]:
# Criando uma função de normalização utilizando o método lower()
def normalizacao(lista_palavras):
    
    lista_normalizada = []
    
    for palavra in lista_palavras:
        
        lista_normalizada.append(palavra.lower())
        
    return lista_normalizada

# Chamando a função e passando os parâmetros
lista_normalizada = normalizacao(lista_palavras)

# Imprimindo uma amostra do resultado
print(lista_normalizada[:10])


['imagem', 'temos', 'a', 'seguinte', 'classe', 'que', 'representa', 'um', 'usuário', 'no']


### Nota/Observação
Agora temos todas as palavras em letras minúsculas, facilitando a normalização

In [14]:
# Imprimindo o total real de palavras existentes em nosso corpus
print(f'O número total de palavras em nosso corpus é de: {len(set(lista_normalizada))}')

O número total de palavras em nosso corpus é de: 18465


# 4. - Preparando estrutura do corretor
Neste passo precisamos ensinar o nosso corretor a identificar erros e desvios de digitação para depois corrigí-los.

A ideia é criar funções capazes de tratar cada tipo de erro identificados no conjunto.

# 4.1 - Tratativa do erro: Falta uma letra

Para este caso vamos criar um exemplo, supondo que ao digitar a palavra **"ortografia"** sem querer enguli uma letra e saiu: **"otografia"**.

Primeiro passo para corrigir e fatiar em duas partes a palavra no lugar exato e inserir a letra faltante entre a lacuna.

In [15]:
# Exemplo

exemplo_palavra = 'otografia'

# Método slicing para fatiar uma string
(exemplo_palavra[:1], exemplo_palavra[1:])

('o', 'tografia')

In [16]:
# Exemplo da função para corrigir
print(f'{exemplo_palavra[:1]} + "a letra que falta (R)" + {exemplo_palavra[1:]}'
    f'\n\nResulta na palavra \n\n{exemplo_palavra[:1] + "r" + exemplo_palavra[1:]}')

o + "a letra que falta (R)" + tografia

Resulta na palavra 

ortografia


Próximo passo é fazer isso para todas as palavras que estiverem faltando uma letra.

In [17]:
# Criando função para inserir as letras

def insere_letras(fatias):
    
    novas_palavras =[]
    
    letras = 'abcedfghijklmnopqrstuvwxyzáâàãéêèíîìóôòõúûùç'
    
    for esquerdo, direito in fatias:
        
        for letra in letras:
            
            novas_palavras.append(esquerdo + letra + direito)
    
    return novas_palavras

# Chamando a função e passando os parâmetros
# Imprimindo uma amostra do resultado
insere_letras([('o', 'tografia')])[:5]

['oatografia', 'obtografia', 'octografia', 'oetografia', 'odtografia']

In [18]:
# Função para gerar as palavras
def gerador_palavras(palavra):
    
    fatias = []
    
    for i in range(len(palavra) + 1):
        
        fatias.append((palavra[:i], palavra[i:]))
    
    palavras_geradas = insere_letras(fatias)
    
    return palavras_geradas

palavras_geradas = gerador_palavras(exemplo_palavra)

# Imprimindo a palavra de exemplo com uma amostra do resultado
print(palavras_geradas[:10])

['aotografia', 'botografia', 'cotografia', 'eotografia', 'dotografia', 'fotografia', 'gotografia', 'hotografia', 'iotografia', 'jotografia']


In [19]:
# Verifica se a palavra correta está dentro da lista gerada
for palavra in palavras_geradas:
    
    if palavra == 'ortografia':
        
        print(f'A palavra correta é "{palavra}" e está dentro da lista gerada!')

A palavra correta é "ortografia" e está dentro da lista gerada!


Após a verificação o corretor precisa saber qual a palavra correta dentro da lista de palavras que foi gerado pela função

Antes de seguir podemos ver a quantidade de palavras geradas anteriormente pela função com o método **len**

In [20]:
print(f'Foram geradas {len(palavras_geradas)} palavras')

Foram geradas 440 palavras


Para está etapa do corretor podemos usar a função **FreqDist()** da lib nltk.

Esse método calcula a frequência que uma determinada palavra em um conjunto dentro do corpus.

Utilizando esse método é preciso passar como input a palavra que desejamos saber a frequência.

In [21]:
# Chamando a função FreqDist
# Passando como parânmetro de input a lista normalizada 
frequencia = nltk.FreqDist(lista_normalizada)

total_palavras = len(lista_normalizada)

# Imprimindo as 10 palavras mais comuns dentro da lista
frequencia.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6368),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

### Nota/Observação
Com este dado, podemos agora calcular a probabilidade de uma determinada palavra aparecer em nosso corpus

Para isso, vamos criar uma função para cálculo da probabilidade

In [22]:
# Criando a função para probabilidade
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada] / total_palavras

In [23]:
# Criando função corretor
def corretor(palavra_errada):
    
    palavras_geradas = gerador_palavras(palavra_errada)
    
    palavra_correta = max(palavras_geradas, key=probabilidade)
    
    # Retorna a palavra já corrigida
    return palavra_correta

Feito a criação das funções iniciais, vamos recriar o exemplo passando a palavra para corrigir.

In [24]:
# Palavra incorreta
exemplo_palavra = 'orografia'

# Chamando a função do corretor
corretor(exemplo_palavra)

'ortografia'

In [25]:
# Teste para a palavra lógica

teste = 'cdigo'
print(f'Você quis dizer: {corretor(teste)}')

Você quis dizer: código


### Nota/Observação
Com os testes executados, podemos observar o corretor já fazendo a primeira tratativa de erro na palavra, próximo passo é verificar a taxa de acerto que conseguimos obter em uma lista de dados.

# 4.1.1 - Taxa de precisão do modelo corretor

Vamos avaliar a qualidade do nosso corretor utilizando o arquivo **palavras.txt**

In [26]:
# Cria função para teste do modelo
def cria_dados_teste(nome_arquivo):
    
    lista_palavras_teste = []
    
    # Realiza a leitura do arquivo txt
    f = open(nome_arquivo, encoding="utf8",mode='r')
    
    for linha in f:
        
        correta, errada = linha.split()
        
        lista_palavras_teste.append((correta, errada))
    
    f.close()
    
    return lista_palavras_teste

# Chama função passando o arquivo
lista_teste = cria_dados_teste('C:\\Users\\batis\\Documents\\Jupyter_Notebook\\palavras.txt')

# Criando função para avaliar o modelo
def avaliador(testes):
    
    numero_palavras = len(testes)
    
    acertou = 0
    
    for correta, errada in testes:
        
        palavra_corrigida = corretor(errada)
        
        # Se a palavra estiver correta incrementa no acerto
        if palavra_corrigida == correta:
            
            acertou += 1
    
    # Calculando a taxa de acerto do modelo
    taxa_acerto = round(acertou * 100 / numero_palavras, 2)
    
    print(f'{taxa_acerto}% de {numero_palavras} palavras')
    

# Chamando a função
avaliador(lista_teste)

1.08% de 186 palavras


Uma performance muito abaixo do desejado acertando cerca de **1.08%** das palavras, isso deve melhorar pois lembrando que estamos tratando somente um erro específico

# 4.2 - Tratativa do erro: Inclusão de uma letra a mais

Para este caso vamos criar um exemplo, supondo que ao digitar a palavra "ortografia" mas quando digitei incluimos um "R" a mais na palavra ficando: **"ortogrrafia"** 

A tratativa para esse erro é a mesma que a anterior, começamos fatiando a palavra no exato ponto e remover a letra em excesso.

In [27]:
# Criando função para remover caractere
def deletando_caracter(fatias):
    
    novas_palavras = []
    
    for esquerdo, direito in fatias:
        
        novas_palavras.append(esquerdo + direito[1:])
    
    return novas_palavras

exemplo = [('ortogr', 'rafia')]

# Chamando a função e passando o parâmetro
deletando_caracter(exemplo)

['ortografia']

Podemos refatorar a função **gerador_palavras** que foi gerada anteriormente, assim tratando os dois tipos de erros.

In [28]:
# Refatorar função gerador_palavras

def gerador_palavras(palavra):
    
    fatias = []
    
    for i in range(len(palavra) + 1):
        
        fatias.append((palavra[:i], palavra[i:]))
    
    palavras_geradas = insere_letras(fatias)
    
    palavras_geradas += deletando_caracter(fatias)
    
    return palavras_geradas

palavra_exemplo = 'ortogrrafia'

palavras_geradas = gerador_palavras(palavra_exemplo)

# Imprimindo uma amostra dos dados
print(palavras_geradas[:5])

['aortogrrafia', 'bortogrrafia', 'cortogrrafia', 'eortogrrafia', 'dortogrrafia']


In [29]:
# Verificar na amostra se contém a palavra correta

for palavra in palavras_geradas:
    
    if palavra == 'ortografia':
        
        print(f'A palavra correta é "{palavra}" e ela está dentro da lista!')

A palavra correta é "ortografia" e ela está dentro da lista!
A palavra correta é "ortografia" e ela está dentro da lista!


In [30]:
# Imprimindo a quantidade de palavras que a função gerador_palavras retornou
print(f'Foram geradas {len(palavras_geradas)} palavras')

Foram geradas 540 palavras


### Nota/Observação
Após refatorar a função adicionando mais uma tratativa de erro, o nosso corretor é capaz de tratar dois tipos de erros apresentados.

Vamos avaliar novamente a taxa de acerto do modelo.

# 4.2.1 - Taxa de precisão do modelo corretor

Vamos avaliar a qualidade do nosso corretor utilizando o arquivo **palavras.txt** novamente.

In [31]:
# Chamando a função que avalia a taxa de acerto do modelo
avaliador(lista_teste)

41.4% de 186 palavras


Obtivemos um aumento considerável na performance de acerto do nosso corretor, subindo para **41.4%**!!

# 4.3 - Tratativa do erro: Troca de letras

Para este caso vamos criar um exemplo, supondo que ao digitar a palavra **"ortografia"** sem querer digito a tecla ao lado formando a palavra: **"ortografoa"** inserindo a letra **"O"** no lugar do **"I"**.

Novamente vamos usar os principios anteriores de fatias a palavra em duas partes e no lugar exato substituir a letra errada pela letra correta.

In [32]:
# Criando função para troca de caractere

def troca_caractere(fatias):
    
    novas_palavras = []
    
    letras = 'abcedfghijklmnopqrstuvwxyzáâàãéêèíîìóôòõúûùç'
    
    for esquerdo, direito in fatias:
        
        for letra in letras:
            
            novas_palavras.append(esquerdo + letra + direito[1:])
        
        return novas_palavras
    
troca_caractere([('ortograf', 'oa')])[:5]

['ortografaa', 'ortografba', 'ortografca', 'ortografea', 'ortografda']

In [33]:
# Refatorar o gerador_palavras incluindo mais uma tratativa de erro

def gerador_palavras(palavra):
    
    fatias = []
    
    for i in range(len(palavra) + 1):
        
        fatias.append((palavra[:i], palavra[i:]))
    
    palavras_geradas = insere_letras(fatias)
    
    palavras_geradas += deletando_caracter(fatias)
    
    palavras_geradas += troca_caractere(fatias)
    
    return palavras_geradas

palavra_exemplo = 'ortografoa'

palavras_geradas = gerador_palavras(palavra_exemplo)

# Imprimindo uma amostra dos dados
print(palavras_geradas[:5])

['aortografoa', 'bortografoa', 'cortografoa', 'eortografoa', 'dortografoa']


In [34]:
# Verificar na amostra se contém a palavra correta

for palavra in palavras_geradas:
    
    if palavra == 'ortografia':
        
        print(f'A palavra correta é "{palavra}" e ela e ela está dentro da lista!')

# 4.4 - Tratativa do erro: Invertendo letras

Para este caso vamos criar um exemplo, supondo que ao digitar a palavra **"ortografia"** sem querer digito a tecla ao lado formando a palavra: **"ortorgrafia"** inserindo a letra **"R"** indevidamente.

Para corrigirmos esse erro, precisamos novamente fatiar em duas partes a palavra no lugar exato mas, dessa vez, trocar as posições das letras de index 0 com a letra de index 1 da fatia direita.

Facilitando nosso trabalho, vou unificar todas as funções em uma única célula para refatorar novamente o gerador_palavras.

In [35]:
# Unificando as funções

#'''-----------------------------------------------------------------------'''
# Função insere letras

def insere_letras(fatias):
    
    novas_palavras =[]
    
    letras = 'abcedfghijklmnopqrstuvwxyzáâàãéêèíîìóôòõúûùç'
    
    for esquerdo, direito in fatias:
        
        for letra in letras:
            
            novas_palavras.append(esquerdo + letra + direito)
        
        return novas_palavras
    
#'''-----------------------------------------------------------------------'''
# Função deleta caracteres

def deletando_caracter(fatias):
    
    novas_palavras = []
    
    for esquerdo, direito in fatias:
        
        novas_palavras.append(esquerdo + direito[1:])
    
    return novas_palavras

#'''-----------------------------------------------------------------------'''
# Função troca caracteres

def troca_caracteres(fatias):
    
    novas_palavras = []
    
    letras = 'abcedfghijklmnopqrstuvwxyzáâàãéêèíîìóôòõúûùç'
    
    for esquerdo, direito in fatias:
        
        for letra in letras:
            
            novas_palavras.append(esquerdo + letra + direito[1:])
    
    return novas_palavras

#'''-----------------------------------------------------------------------'''
# Função invertendo caracteres

def invertendo_caracteres(fatias):
    
    novas_palavras = []
    
    for esquerdo, direito in fatias:
        
        if len(direito) > 1:
            
            novas_palavras.append(esquerdo + direito[1] + direito[0] + direito[2:])
    
    return novas_palavras

#'''-----------------------------------------------------------------------'''
# Função principal para gerar as palavras
# Função invertendo caracteres

def gerador_palavras(palavra):
    
    fatias = []
    
    for i in range(len(palavra) + 1):
        
        fatias.append((palavra[:i], palavra[i:]))
        
    palavras_geradas = insere_letras(fatias)
    
    palavras_geradas += deletando_caracter(fatias)
    
    palavras_geradas += troca_caracteres(fatias)
    
    palavras_geradas += invertendo_caracteres(fatias)
    
    return palavras_geradas

In [36]:
# Teste do modelo

palavra_exemplo = 'ortorgrafia'

palavras_geradas = gerador_palavras(palavra_exemplo)